File: 算法笔记/NQSDemo/src/main/java/jeason/b树形结构/a二叉树/Printer工具类/base/Printer.java
package jeason.b树形结构.a二叉树.Printer工具类.base;

public abstract class Printer {	
	/**
	 * 二叉树的基本信息
	 */
	protected BinaryTreeInfoProtocol tree;
	
	public Printer(BinaryTreeInfoProtocol tree) {
		this.tree = tree;
	}
	
	/**
	 * 生成打印的字符串
	 */
	public abstract String printString();
	
	/**
	 * 打印后换行
	 */
	public void println() {
		print();
		System.out.println();
	}
	
	/**
	 * 打印
	 */
	public void print() {
		System.out.print(printString());
	}
}

File: PythonAlgorithm/Study/b树形结构/a二叉搜索树/Printer工具类/Printer.py
from abc import ABC, abstractmethod
from .BinaryTree_AbstractInfo import BinaryTree_AbstractInfo


class Printer(ABC):
    """
    抽象打印器类,负责按不同顺序打印二叉树
    """
    def __init__(self, tree: BinaryTree_AbstractInfo):
        self.tree: BinaryTree_AbstractInfo = tree

    @abstractmethod
    def print_string(self) -> str:
        """生成打印字符串"""
        # 抽象方法，没有实现

    def println(self):
        """打印后换行"""
        self.print()
        print()

    def print(self):
        """打印"""
        print(self.print_string())


